README.md
<p align="center">
  <a href="https://shweshi.github.io/OpenGraph/">
    <img alt="Laravel OpenGraph" src="https://user-images.githubusercontent.com/35127382/56821996-695df500-686d-11e9-9c59-c70d0c78c80e.png" />
  </a>
</p>

<p align="center">
  <strong>A Laravel package to fetch Open Graph metadata of a website.</strong>
</p>

<p align="center">
  <a href="https://packagist.org/packages/shweshi/OpenGraph">
    <img src="https://img.shields.io/packagist/v/shweshi/OpenGraph.svg?style=flat-square" alt="Latest Version on Packagist">
  </a>

  <a href="https://packagist.org/packages/shweshi/OpenGraph">
    <img src="https://img.shields.io/packagist/dt/shweshi/OpenGraph.svg?style=flat-square" alt="Total Downloads">
  </a>

   <a href="https://scrutinizer-ci.com/g/shweshi/OpenGraph/?branch=master">
      <img src="https://scrutinizer-ci.com/g/shweshi/OpenGraph/badges/quality-score.png?b=master" alt="Scrutinizer Code Quality">
  </a>

  <a href="https://styleci.io/repos/116995669">
      <img src="https://styleci.io/repos/116995669/shield?branch=master" alt="StyleCI">
  </a>

  <a href="https://scrutinizer-ci.com/g/shweshi/OpenGraph/build-status/master">
    <img src="https://scrutinizer-ci.com/g/shweshi/OpenGraph/badges/build.png?b=master" alt="Build Status">
  </a>

  <a href="https://app.fossa.io/projects/git%2Bgithub.com%2Fshweshi%2FOpenGraph?ref=badge_shield">
    <img src="https://app.fossa.io/api/projects/git%2Bgithub.com%2Fshweshi%2FOpenGraph.svg?type=shield" alt="FOSSA Status">
  </a>
</p>

## Features

- **Easily fetch metadata of a url.** Laravel OpenGraph fetch all the metadata of a URL.

- **Supports language specific metadata.** Laravel OpenGraph can fetch metadata in a specific language if webpage supports.

- **Supports twitter metadata.** Laravel OpenGraph supports twitter OG data too.

- **Verify image URL.** Laravel OpenGraph verifies that the image url in the image metadata is valid or not.

## Demo

```
  curl https://laravelopengraph.herokuapp.com/api/fetch?url=ogp.me&allMeta=true&language=en_GB
```

## How to use Laravel OpenGraph

Article can be found on medium blog: https://hackernoon.com/how-to-fetch-open-graph-metadata-in-laravel-2d5d674904d7

### Installation

Perform the following operations in order to use this package

- Install via composer

```
composer require "quangtam/opengraph"
```

If you do not run Laravel 5.5 (or higher), then add the service provider in config/app.php:

- **Add Service Provider**
  Open `config/app.php` and add `shweshi\OpenGraph\Providers\OpenGraphProvider::class,` to the end of `providers` array:

  ```
  'providers' => array(
      ....
      quangtam\OpenGraph\Providers\OpenGraphProvider::class,
  ),
  ```

  Next under the `aliases` array:

  ```
  'aliases' => array(
      ....
      'OGMeta' => quangtam\OpenGraph\Facades\OpenGraphFacade::class
  ),
  ```

If you do run the package on Laravel 5.5+, package auto-discovery takes care of the magic of adding the service provider.

### Requirements

- You need to install the [DOM](http://www.php.net/en/dom) extension.

### How to use

- After following the above steps,

  ```
  use OGMeta;

  $data = OGMeta::fetch("https://unsplash.com/");
  ```

  this will give you an array like this..

  ```
    array (
      'title' => 'Beautiful Free Images & Pictures | Unsplash',
      'description' => 'Beautiful, free images and photos that you can download and use for any project. Better than any royalty free or stock photos.',
      'type' => 'website',
      'url' => 'https://unsplash.com/',
      'image' => 'http://images.unsplash.com/photo-1542841791-1925b02a2bbb?ixlib=rb-0.3.5&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjEyMDd9&s=aceabe8a2fd1a273da24e68c21768de0',
      'image:secure_url' => 'https://images.unsplash.com/photo-1542841791-1925b02a2bbb?ixlib=rb-0.3.5&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjEyMDd9&s=aceabe8a2fd1a273da24e68c21768de0',
    )
  ```

  You can also pass an optional parameter either true or false along with url. When set false it will only fetch basic metadata and in case of true it will fetch all the other optional metadata as well like audio, video, music and twitter metatags as well.

  ```
  $data = OGMeta::fetch("https://unsplash.com/", true);
  ```

  this will give you an array like this..

  ```
    array (
      'charset' => 'UTF8',
      'viewport' => 'width=device-width, initial-scale=1.0, maximum-scale=1.0, minimal-ui',
      'mobile-web-app-capable' => 'yes',
      'apple-mobile-web-app-capable' => 'yes',
      'apple-mobile-web-app-title' => 'Unsplash',
      'application-name' => 'Unsplash',
      'author' => 'Unsplash',
      'msapplication-config' => 'browserconfig.xml',
      'msapplication-TileColor' => '#ffffff',
      'msapplication-TileImage' => 'https://unsplash.com/mstile-144x144.png',
      'theme-color' => '#ffffff',
      'description' => 'Beautiful, free images and photos that you can download and use for any project. Better than any royalty free or stock photos.',
      'twitter:site' => '@unsplash',
      'twitter:title' => 'Beautiful Free Images & Pictures | Unsplash',
      'twitter:description' => 'Beautiful, free images and photos that you can download and use for any project. Better than any royalty free or stock photos.',
      'twitter:url' => 'https://unsplash.com/',
      'twitter:card' => 'summary_large_image',
      'twitter:image' => 'https://images.unsplash.com/photo-1546486610-e9fe4f1e6751?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjEyMDd9',
      'title' => 'Beautiful Free Images & Pictures | Unsplash',
      'type' => 'website',
      'url' => 'https://unsplash.com/',
      'image' => 'http://images.unsplash.com/photo-1546486610-e9fe4f1e6751?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjEyMDd9',
      'image:secure_url' => 'https://images.unsplash.com/photo-1546486610-e9fe4f1e6751?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjEyMDd9',
  )
  ```

  To fetch the metadata in a specific language you can pass the language as the third argument, this value will be used as the Accept-Language header.

  ```
  $url = "https://ogp.me",
  $allMeta = true, // can be false
  $language = 'en' // en-US,en;q=0.8,en-GB;q=0.6,es;q=0.4
  $data = OpenGraph::fetch($url, $allMeta, $language);
  ```

### Testing

    composer test

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

### Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE OF CONDUCT](CODE_OF_CONDUCT.md) for details.

### License

The MIT License (MIT). Please see [License File](LICENSE) for more information.

[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fshweshi%2FOpenGraph.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2Fshweshi%2FOpenGraph?ref=badge_large)

### Happy coding!


CHANGELOG.md
# Changelog


All notable changes to `OpenGraph` will be documented in this file.

## 1.0.6 - 2019-06-30

- Fix conflict with SEO Meta Tool of Laravel
- Add method to get meta from html content OGMeta::get()

## 1.0.5 - 2019-06-13

- Added support to fetch language specific metadata.

## 1.0.4 - 2019-04-26

- Verify that the image url is valid in the metadata.

## 1.0.3 - 2019-01-04

- Added support to fetch all metadata

## 1.0.2 - 2018-11-23

- Added tests and autoload for laravel >= 5.5

## 1.0.1 - 2018-01-17

- Fix autoload

## 1.0.0 - 2018-01-17

- initial release


src/OpenGraph.php
<?php

namespace quangtam\OpenGraph;

use DOMDocument;

class OpenGraph
{
    public function fetch($url, $allMeta = null, $lang = null)
    {
        $html = $this->curl_get_contents($url, $lang);
        /**
         * parsing starts here:.
         */
        return $this->get($html, $allMeta, $lang);
    }

    public function get($html, $allMeta = null, $lang = null)
    {
        /**
         * parsing starts here:.
         */
        $doc = new DOMDocument();
        @$doc->loadHTML('<?xml encoding="utf-8" ?>'.$html);

        $tags = $doc->getElementsByTagName('meta');
        $metadata = [];
        foreach ($tags as $tag) {
            $metaproperty = ($tag->hasAttribute('property')) ? $tag->getAttribute('property') : $tag->getAttribute('name');
            if (!$allMeta && $metaproperty && strpos($tag->getAttribute('property'), 'og:') === 0) {
                $key = strtr(substr($metaproperty, 3), '-', '_');
                $value = $tag->getAttribute('content');
            }
            if ($allMeta && $metaproperty) {
                $key = (strpos($metaproperty, 'og:') === 0) ? strtr(substr($metaproperty, 3), '-', '_') : $metaproperty;
                $value = $tag->getAttribute('content');
            }
            if (!empty($key)) {
                $metadata[$key] = $value;
            }
            /*
             * Verify image url
             */
            if (isset($metadata['image'])) {
                $isValidImageUrl = $this->verify_image_url($metadata['image']);
                if (!$isValidImageUrl) {
                    $metadata['image'] = '';
                }
            }
        }

        return $metadata;
    }

    public function getJsonLd($html) {
        $doc = new DOMDocument();
        @$doc->loadHTML('<?xml encoding="utf-8" ?>'.$html);

        $tags = $doc->getElementsByTagName('script');

        foreach ($tags as $tag) {
            $metaproperty = ($tag->hasAttribute('type')) ? $tag->getAttribute('type') : '';
            if($metaproperty == 'application/ld+json') {
                return $tag;
            }
        }
    }

    protected function curl_get_contents($url, $lang)
    {
        $headers = [
          'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
          'Cache-Control: no-cache',
          'User-Agent: Curl',
        ];

        if ($lang) {
            array_push($headers, 'Accept-Language: '.$lang);
        }

        $curl = curl_init();

        curl_setopt_array($curl, [
          CURLOPT_URL            => $url,
          CURLOPT_FAILONERROR    => false,
          CURLOPT_FOLLOWLOCATION => true,
          CURLOPT_RETURNTRANSFER => true,
          CURLOPT_SSL_VERIFYHOST => false,
          CURLOPT_SSL_VERIFYPEER => false,
          CURLOPT_ENCODING       => 'UTF-8',
          CURLOPT_MAXREDIRS      => 10,
          CURLOPT_TIMEOUT        => 30,
          CURLOPT_HTTP_VERSION   => CURL_HTTP_VERSION_1_1,
          CURLOPT_CUSTOMREQUEST  => 'GET',
          CURLOPT_HTTPHEADER     => $headers,
        ]);

        $response = curl_exec($curl);
        curl_close($curl);

        return $response;
    }

    protected function verify_image_url($url)
    {
        $headers = get_headers($url);

        return stripos($headers[0], '200 OK') ? true : false;
    }
}
